### PR TITLE
Ftr: etcdv3 for registry 

### PR DESCRIPTION
**Pull request by sxllwx**
**What this PR does**:

Ftr: etcdv3 for registry 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

remoting pkg unit test is done, see ./remoting/etcdv3/*_test.go

registry pkg unit test is done, see ./registry/etcdv3/*_test.go

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: registry/etcdv3/listener_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
+package etcdv3
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/apache/dubbo-go/common"
```
Comment (fangyincheng) [inline registry/etcdv3/listener_test.go]: format import

File: registry/etcdv3/registry.go [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/common/utils"
+	"github.com/apache/dubbo-go/registry"
+	"github.com/apache/dubbo-go/remoting/etcdv3"
+	"github.com/apache/dubbo-go/version"
+)
+
+import (
+	"github.com/juju/errors"
```
Comment (fangyincheng) [inline registry/etcdv3/registry.go]: We agreed to use only `perrors`.  Is it `github.com/juju/errors`  necessary ？

File: remoting/etcdv3/client_test.go [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/suite"
+
+	"github.com/coreos/etcd/mvcc/mvccpb"
+	"github.com/juju/errors"
+	"google.golang.org/grpc/connectivity"
```
Comment (fangyincheng) [inline remoting/etcdv3/client_test.go]: format  import and don't use juju/errors。

File: registry/etcdv3/listener_test.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	cfg := embed.NewConfig()
+	cfg.Dir = "/tmp/default.etcd"
+	e, err := embed.StartEtcd(cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	select {
+	case <-e.Server.ReadyNotify():
+		t.Log("Server is ready!")
+	case <-time.After(60 * time.Second):
```
Comment (AlexStocks) [inline registry/etcdv3/listener_test.go]: use getty.GetTimeWheel().After instead

File: registry/etcdv3/listener_test.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	if !listener.DataChange(remoting.Event{Path: "/dubbo/com.ikurento.user.UserProvider/providers/jsonrpc%3A%2F%2F127.0.0.1%3A20001%2Fcom.ikurento.user.UserProvider%3Fanyhost%3Dtrue%26app.version%3D0.0.1%26application%3DBDTService%26category%3Dproviders%26cluster%3Dfailover%26dubbo%3Ddubbo-provider-golang-2.6.0%26environment%3Ddev%26group%3D%26interface%3Dcom.ikurento.user.UserProvider%26ip%3D10.32.20.124%26loadbalance%3Drandom%26methods.GetUser.loadbalance%3Drandom%26methods.GetUser.retries%3D1%26methods.GetUser.weight%3D0%26module%3Ddubbogo%2Buser-info%2Bserver%26name%3DBDTService%26organization%3Dikurento.com%26owner%3DZX%26pid%3D74500%26retries%3D0%26service.filter%3Decho%26side%3Dprovider%26timestamp%3D1560155407%26version%3D%26warmup%3D100"}) {
+		t.Fatal("data change not ok")
+	}
+}
+
+func TestRegistrySuite(t *testing.T) {
+	suite.Run(t, &RegistryTestSuite{})
+}
+
+type MockDataListener struct {
```
Comment (AlexStocks) [inline registry/etcdv3/listener_test.go]: write code in 1 line if the struct body is empty

File: registry/etcdv3/listener_test.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func (*MockDataListener) Process(configType *remoting.ConfigChangeEvent) {
+
```
Comment (AlexStocks) [inline registry/etcdv3/listener_test.go]: the same as above

File: remoting/etcdv3/facade.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+		case <-r.Client().Done():
+			r.ClientLock().Lock()
+			clientName := RegistryETCDV3Client
+			timeout, _ := time.ParseDuration(r.GetUrl().GetParam(constant.REGISTRY_TIMEOUT_KEY, constant.DEFAULT_REG_TIMEOUT))
+			endpoint := r.GetUrl().Location
+			r.Client().Close()
+			r.SetClient(nil)
+			r.ClientLock().Unlock()
+
+			// 接zk，直至成功
```
Comment (AlexStocks) [inline remoting/etcdv3/facade.go]: connect zk?

File: remoting/etcdv3/facede_test.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
+package etcdv3
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/apache/dubbo-go/common"
```
Comment (AlexStocks) [inline remoting/etcdv3/facede_test.go]: another import block

File: remoting/etcdv3/listener_test.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
+package etcdv3
+
+import (
+	"time"
+
+	"github.com/apache/dubbo-go/remoting"
```
Comment (AlexStocks) [inline remoting/etcdv3/listener_test.go]: another 2 imports pls.

File: remoting/etcdv3/listener.go [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+			}
+		}
+		return false
+	case mvccpb.DELETE:
+		logger.Warnf("etcd get event (key{%s}) = event{EventNodeDeleted}", event.Kv.Key)
+		return true
+	}
+
+	panic("unreachable")
+}
```
Comment (hxmhlt) [inline remoting/etcdv3/listener.go]: why panic?

File: registry/etcdv3/registry.go (hunk 2 above)
Comment (AlexStocks) [reply registry/etcdv3/registry.go]: @fangyincheng agree with u.

File: remoting/etcdv3/client_test.go (hunk 3 above)
Comment (AlexStocks) [reply remoting/etcdv3/client_test.go]: yeah.

File: remoting/etcdv3/facade.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+	defer r.WaitGroup().Done()
+LOOP:
+	for {
+		select {
+		case <-r.GetDone():
+			logger.Warnf("(ETCDV3ProviderRegistry)reconnectETCDV3 goroutine exit now...")
+			break LOOP
+			// re-register all services
+		case <-r.Client().Done():
+			r.ClientLock().Lock()
```
Comment (hxmhlt) [inline remoting/etcdv3/facade.go]: This case will be triggered when client.done is closed.  So where will the client.done be closed?

File: registry/etcdv3/listener.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/registry"
+	"github.com/apache/dubbo-go/remoting"
+)
+
+import (
+	"github.com/juju/errors"
```
Comment (fangyincheng) [inline registry/etcdv3/listener.go]: don't use juju/errors

File: remoting/etcdv3/client.go [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+import (
+	"context"
+	"path"
+	"sync"
+	"time"
+)
+import (
+	"github.com/apache/dubbo-go/common/logger"
+)
+import (
```
Comment (fangyincheng) [inline remoting/etcdv3/client.go]: this import block should be second.

File: remoting/etcdv3/facade.go [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+	"time"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/logger"
+)
+
+import (
```
Comment (fangyincheng) [inline remoting/etcdv3/facade.go]: this import block should be in second.

File: remoting/etcdv3/listener.go [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+	"sync"
+	"time"
+)
+
+import (
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/remoting"
+)
+
+import (
```
Comment (fangyincheng) [inline remoting/etcdv3/listener.go]: this import block should be in second.

File: remoting/etcdv3/listener.go (hunk 10 above)
Comment (sxllwx) [reply remoting/etcdv3/listener.go]: The etcd package will only return two types of event types, PUT and DELETE. It is impossible to have a third type. I have already dealt with both cases. Golang requires that the switch statement must be default if you want to end the entire function. And I don't have this kind of scene here. So the last sentence in the code uses a panic.

File: remoting/etcdv3/listener.go [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+	logger.Infof("listen dubbo provider key{%s} event and wait to get all provider etcdv3 nodes", key)
+	go func(key string, listener remoting.DataListener) {
+		l.ListenServiceNodeEventWithPrefix(key, listener)
+		logger.Warnf("listenDirEvent(key{%s}) goroutine exit now", key)
+	}(key, listener)
+
+	logger.Infof("listen dubbo service key{%s}", key)
+	go func(key string) {
+		if l.ListenServiceNodeEvent(key) {
+			listener.DataChange(remoting.Event{Path: key, Action: remoting.EventTypeDel})
```
[the remark this answers is not in the document]
Comment (sxllwx) [reply remoting/etcdv3/listener.go]: Yes.  If etcd client be closed , ListenServiceNodeEvent will return false

File: remoting/etcdv3/facade.go (hunk 11 above)
Comment (sxllwx) [reply remoting/etcdv3/facade.go]: etcdV3Registry.Destroy() -> remoting/etcdv3.Client.Close() ->  remoting/etcdv3.Client.stop()

Comment (hxmhlt) [reply remoting/etcdv3/facade.go]: Means only call destory method can trigger reconnect? If the connection of etcd is broken, how can it be reconnected?

Comment (hxmhlt) [reply remoting/etcdv3/facade.go]: I have not found the etcdV3Registry.Destroy()function be called except in unit_test.

File: remoting/etcdv3/client_test.go [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+	"reflect"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+)
+
+import (
+	"github.com/coreos/etcd/mvcc/mvccpb"
+	"github.com/pkg/errors"
```
Comment (fangyincheng) [inline remoting/etcdv3/client_test.go]: It has the same name as errors in the standard library. So we need rename to "perrors", eg: perrors "github.com/pkg/errors".   
And check all please.

File: registry/etcdv3/listener.go [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func (l *dataListener) AddInterestedURL(url *common.URL) {
+	l.interestedURL = append(l.interestedURL, url)
+}
+
+func (l *dataListener) DataChange(eventType remoting.Event) bool {
+
+	url := eventType.Path[strings.Index(eventType.Path, "/providers/")+len("/providers/"):]
+	serviceURL, err := common.NewURL(context.TODO(), url)
```
Comment (AlexStocks) [inline registry/etcdv3/listener.go]: In dubbogo, we always use context.Background().

File: registry/etcdv3/listener.go [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+
+func (l *configurationListener) Next() (*registry.ServiceEvent, error) {
+	for {
+		select {
+		case <-l.registry.done:
+			logger.Warnf("listener's etcd client connection is broken, so etcd event listener exit now.")
+			return nil, errors.New("listener stopped")
+
+		case e := <-l.events:
+			logger.Warnf("got etcd event %#s", e)
```
Comment (AlexStocks) [inline registry/etcdv3/listener.go]: maybe logger.Info be better

File: registry/etcdv3/registry.go [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+
+	timeout, err := time.ParseDuration(url.GetParam(constant.REGISTRY_TIMEOUT_KEY, constant.DEFAULT_REG_TIMEOUT))
+	if err != nil {
+		logger.Errorf("timeout config %v is invalid ,err is %v",
+			url.GetParam(constant.REGISTRY_TIMEOUT_KEY, constant.DEFAULT_REG_TIMEOUT), err.Error())
+		return nil, errors.WithMessagef(err, "new etcd registry(address:%+v)", url.Location)
+	}
+
+	logger.Infof("etcd address is: %v", url.Location)
+	logger.Infof("time-out is: %v", timeout.String())
```
Comment (AlexStocks) [inline registry/etcdv3/registry.go]: this is nonsense. u should use the same log.

File: registry/etcdv3/registry.go [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+func (r *etcdV3Registry) Register(svc common.URL) error {
+
+	role, err := strconv.Atoi(r.URL.GetParam(constant.ROLE_KEY, ""))
+	if err != nil {
+		return errors.WithMessage(err, "get registry role")
+	}
+
+	r.cltLock.Lock()
+	if _, ok := r.services[svc.Key()]; ok {
+		return errors.New(fmt.Sprintf("Path{%s} has been registered", svc.Path))
```
Comment (AlexStocks) [inline registry/etcdv3/registry.go]: deadlock?

File: registry/etcdv3/registry.go [hunk 22]
```diff
@@ -0,0 +1,323 @@
...
+	params.Add("dubbo", "dubbo-provider-golang-"+version.Version)
+	params.Add("side", (common.RoleType(common.PROVIDER)).Role())
+
+	if len(svc.Methods) == 0 {
+		params.Add("methods", strings.Join(svc.Methods, ","))
+	}
+
+	logger.Debugf("provider url params:%#v", params)
+	var host string
+	if svc.Ip == "" {
```
Comment (AlexStocks) [inline registry/etcdv3/registry.go]: len(svc.Ip) == 0

File: remoting/etcdv3/listener.go [hunk 23]
```diff
@@ -0,0 +1,219 @@
...
+
+	logger.Infof("get key children list %s, keys %v values %v", key, keyList, valueList)
+
+	for i, k := range keyList {
+		logger.Infof("got children list key -> %s", k)
+		if !listener.DataChange(remoting.Event{
+			Path:    k,
+			Action:  remoting.EventTypeAdd,
+			Content: valueList[i],
+		}) {
```
Comment (AlexStocks) [inline remoting/etcdv3/listener.go]: this check is very smelly. the right way should be as follows.

res := listener.Datachagen()
if !res {}

File: registry/etcdv3/listener_test.go (hunk 1 above)
Comment (sxllwx) [reply registry/etcdv3/listener_test.go]: Thx, i had reformat import, in latest commit.

File: registry/etcdv3/registry.go (hunk 2 above)
Comment (sxllwx) [reply registry/etcdv3/registry.go]: Thx, i had change github.comjuju/errors to  github.com/pkg/errors

File: registry/etcdv3/listener_test.go (hunk 4 above)
Comment (sxllwx) [reply registry/etcdv3/listener_test.go]: Thx, i had change to  getty.GetTimeWheel().After in latest commit.

File: registry/etcdv3/listener_test.go (hunk 5 above)
Comment (sxllwx) [reply registry/etcdv3/listener_test.go]: Thx, i had reformat empty struct in latest commit

File: registry/etcdv3/listener_test.go (hunk 6 above)
Comment (sxllwx) [reply registry/etcdv3/listener_test.go]: Thx, i had reformat empty function in latest commit

File: remoting/etcdv3/facede_test.go (hunk 8 above)
Comment (sxllwx) [reply remoting/etcdv3/facede_test.go]: Thx, i have reformat import in latest commit.

File: registry/etcdv3/listener.go (hunk 12 above)
Comment (sxllwx) [reply registry/etcdv3/listener.go]: Thx, i had change github.comjuju/errors to github.com/pkg/errors

File: remoting/etcdv3/client.go (hunk 13 above)
Comment (sxllwx) [reply remoting/etcdv3/client.go]: Thx, i had reformat import block in latest commit.

File: remoting/etcdv3/facade.go (hunk 14 above)
Comment (sxllwx) [reply remoting/etcdv3/facade.go]: Thx, i had reformat import block in latest commit.

File: remoting/etcdv3/listener.go (hunk 15 above)
Comment (sxllwx) [reply remoting/etcdv3/listener.go]: Thx, i had reformat import block in latest commit.

File: registry/etcdv3/registry.go [hunk 24]
```diff
@@ -0,0 +1,323 @@
...
+
+var (
+	processID = ""
+	localIP   = ""
+)
+
+func init() {
+	processID = fmt.Sprintf("%d", os.Getpid())
+	localIP, _ = utils.GetLocalIP()
+	extension.SetRegistry("etcdv3", newETCDV3Registry)
```
[the remark this answers is not in the document]
Comment (sxllwx) [reply registry/etcdv3/registry.go]: Thx, i had add constant  named 'Name' in registry/etcdv3/ pkg.

File: registry/etcdv3/registry.go [hunk 25]
```diff
@@ -0,0 +1,323 @@
...
+
+		// new client & listener
+		listener := etcdv3.NewEventListener(r.client)
+
+		r.listenerLock.Lock()
+		r.listener = listener
+		r.listenerLock.Unlock()
+	}
+
+	//注册到dataconfig的interested
```
[the remark this answers is not in the document]
Comment (sxllwx) [reply registry/etcdv3/registry.go]: Thx, i had translate into english.

File: registry/etcdv3/registry_test.go [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+	"testing"
+	"time"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+)
+
+import (
```
[the remark this answers is not in the document]
Comment (sxllwx) [reply registry/etcdv3/registry_test.go]: Thx, i had reformat import block in latest commit.

File: registry/etcdv3/listener.go (hunk 18 above)
Comment (sxllwx) [reply registry/etcdv3/listener.go]: Thx, i had change context.TODO() to context.Backgroud().

File: registry/etcdv3/registry.go (hunk 20 above)
Comment (sxllwx) [reply registry/etcdv3/registry.go]: Thx, i had rebuild the logoutput

File: registry/etcdv3/listener.go (hunk 19 above)
Comment (sxllwx) [reply registry/etcdv3/listener.go]: Yes, it should be info, warn will print stack~

File: remoting/etcdv3/client_test.go (hunk 17 above)
Comment (sxllwx) [reply remoting/etcdv3/client_test.go]: Thx, i had rename github.com/pkg/errors to perrors.

File: registry/etcdv3/registry.go (hunk 21 above)
Comment (sxllwx) [reply registry/etcdv3/registry.go]: Yes,  this is a serious bug, should unlock before return.

File: remoting/etcdv3/listener.go (hunk 23 above)
Comment (sxllwx) [reply remoting/etcdv3/listener.go]: Yes, i had fix this format in latest commit.

File: registry/etcdv3/registry.go (hunk 22 above)
Comment (sxllwx) [reply registry/etcdv3/registry.go]: Yes, len(string), will more efficient.

File: remoting/etcdv3/client_test.go (hunk 3 above)
Comment (sxllwx) [reply remoting/etcdv3/client_test.go]: Thx, i had reformat all import && change juju/errors to  pkg/errors.

File: remoting/etcdv3/facade.go (hunk 7 above)
Comment (sxllwx) [reply remoting/etcdv3/facade.go]: Thx,  i had fix this mis spell in latest commit.

File: remoting/etcdv3/listener_test.go (hunk 9 above)
Comment (sxllwx) [reply remoting/etcdv3/listener_test.go]: Thx, i have reformat import in latest commit.

File: remoting/etcdv3/facade.go (hunk 11 above)
Comment (sxllwx) [reply remoting/etcdv3/facade.go]: Thx,  i had close remoting/etcdv3/client.Client.ctx. when etcd connection lose.

**Comment by AlexStocks** [comment]
LGTM

File: remoting/etcdv3/facede_test.go [hunk 27]
```diff
@@ -0,0 +1,71 @@
+package etcdv3
```
Comment (hxmhlt) [inline remoting/etcdv3/facede_test.go]: wrong file name not facede but facade

Comment (sxllwx) [reply remoting/etcdv3/facede_test.go]: Thx, i had rename testfile.

**Pull request merged**